File: src/main/java/core/procedures/sets/Difference.java
package core.procedures.sets;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.HashSet;
import java.util.Set;

public final class Difference extends AFn {

  public Difference() {
    super(new FnArgsBuilder().minArgs(1).mandatoryArgsTypes(new Class[]{Set.class}).restArgsType(Set.class));
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "difference";
  }

  @Override
  public Set<Object> apply(Object... args) {
    if (args.length == 1) {
      return (Set<Object>) args[0];
    }
    Set result = new HashSet((Set)args[0]);
    for (int i = 1, argsLength = args.length; i < argsLength; i++) {
      result.removeAll((Set)args[i]);
    }
    return result;
  }
}

File: src/main/java/core/procedures/sets/Difference.kt
package core.procedures.sets

import core.procedures.AFn
import core.procedures.Arity.AtLeast
import core.scm.MutableSet

class Difference : AFn<Any?, Set<*>>(name = "difference", isPure = true, arity = AtLeast(1),
                                     mandatoryArgsTypes = arrayOf(Set::class.java), restArgsType = Set::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        1    -> args[0]!! as Set<*>
        else -> MutableSet(args[0]!! as Set<*>).apply {
            for (i in (1 until args.size)) {
                removeAll(args[i]!! as Set<*>)
            }
        }
    }
}
